### PR TITLE
add batch PointOp verification

### LINKED ISSUE #11: Batch PointOp verification
Point ops need to be batch-verified for efficiency. This is how it's done:

If one has two statements `x·A + y·B == 0` and `z·C + w·D == 0` (number of terms could vary), they can be combined together with a free variable `e`:

```
x·A + y·B == 0
+
e·(z·C + w·D == 0)
=
x·A + y·B + e·z·C + e·w·D == 0
```

If the latter statement is true for all `e`, then it implies that each individual of first two statements is also true. The free variable `e` will be sampled randomly via `Scalar::random(&mut rand::thread_rng())` to make it completely non-deterministic (unlike fiat-shamir challenges the prover does not ever need to know `e` here, and _must not_ know it to prevent malicious cancellation of some terms in one statement by terms in another).

The inexpensive scalar-scalar multiplications are precomputed (`e·z` and `e·w`) and then the resulting arrays of scalar factors and points are sent into `RistrettoPoint::optional_multiscalar_mul`.

```
RistrettoPoint::optional_multiscalar_mul(
  [x, y, e·z, e·w],
  [A, B, C, D].map(|p|p.decompress()),
)
```

Note 1: the `optional_multiscalar_mul` takes iterators of `Option<RistrettoPoint>` to handle decompression errors in the inlined implementation and spare the user from allocating an intermediate vector.
Note 2: the `optional_multiscalar_mul` takes _iterators_ instead of slices or vectors, so that the user can pass a combination of `.map`s that do premultiplication by `e` w/o ever having to `.collect()` the iterator into a container.

How does this scale to N point ops? Easy: just generate a random `e[i]` for each `i`th statement (including the first one to avoid special cases). Premultiply all scalars with `e[i]` in the `i`th statement and then pass the concatenated arrays into `multiscalar_mul`.

Since the number of operations is not static, we'd have to allocate intermediate buffer anyway (could not keep them as chained-iterator), but we can do it once (once per scalars, once per points) using `Vec::with_capacity()` using the sum of lengths of all terms from all statements.

For reference, there's an old PR on Bulletproofs (https://github.com/dalek-cryptography/bulletproofs/pull/86) where batching is done for rangeproofs. The code is the easiest to see here: https://github.com/dalek-cryptography/bulletproofs/blob/oleg/batch-verify/src/range_proof/mod.rs (note use of `Verification` object similar to `PointOp` and `batch_challenge`s that protect statements from cancelling each other maliciously).

### PR DESCRIPTION
**Pull request by vickiniu**
Addresses #11, adds batch PointOp verification

### CODE REVIEW

File: src/point_ops.rs [hunk 1]
```diff
@@ -51,7 +51,47 @@ impl PointOp {
...
+
+        let length: usize = batch.iter().map(|p| p.arbitrary.len()).sum();
+
+        let mut weights: Vec<Scalar> = Vec::with_capacity(length + 2);
+        let mut points: Vec<Option<RistrettoPoint>> = Vec::with_capacity(length + 2);
+
+        // Add base points
+        points.push(Some(gens.B));
+        points.push(Some(gens.B_blinding));
+        weights.push(Scalar::default());
```
Comment (oleganza) [inline src/point_ops.rs]: Nitpick: clarity.
```suggestion
        weights.push(Scalar::zero());
```

File: src/point_ops.rs [hunk 2]
```diff
@@ -51,7 +51,47 @@ impl PointOp {
...
+        weights.push(Scalar::default());
```
Comment (oleganza) [inline src/point_ops.rs]: Same
```suggestion
        weights.push(Scalar::zero());
```

File: src/point_ops.rs (hunk 1 above)
Comment (vickiniu) [reply src/point_ops.rs]: ah that is much better, thanks!

File: src/point_ops.rs [hunk 3]
```diff
@@ -51,7 +51,47 @@ impl PointOp {
...
+
+        // Iterate over every point, adding both weights and points to
+        // our arrays
+        for p in batch.iter() {
+            // Sample free variable e
+            let e = Scalar::random(&mut rand::thread_rng());
```
Comment (oleganza) [inline src/point_ops.rs]: this creates a new RNG instance in each iteration. I'd move it out of the loop and reuse the RNG:

```
let mut rng = rand::thread_rng();
...
let e = Scalar::random(&mut rng);
```

**Pull request merged**